python/ray/tests/test_object_assign_owner.py

```diff
@@ -0,0 +1,77 @@
+import pytest
+import ray
+import time
+
+
+@pytest.mark.parametrize(
+    "actor_resources",
+    [
+        dict(
+            zip(["owner", "creator", "borrower"], [{
+                f"node{i}": 1
+            } for i in _])) for _ in [
+                [1, 2, 3],  # None of them is on the same node.
+                [1, 1, 3],  # Owner and creator are on the same node.
+                [3, 2, 3],  # Owner and borrower are on the same node.
+                [1, 3, 3],  # Creator and borrower are on the same node.
+                [3, 3, 3],  # All of them are on the same node.
+            ]
+    ])
+def test_owner_assign_when_put(ray_start_cluster, actor_resources):
+    cluster_node_config = [{
+        "num_cpus": 1,
+        "resources": {
+            f"node{i+1}": 10
+        }
+    } for i in range(3)]
+    cluster = ray_start_cluster
+    for kwargs in cluster_node_config:
+        cluster.add_node(**kwargs)
+    ray.init(address=cluster.address)
+
+    @ray.remote(resources=actor_resources["creator"], num_cpus=0)
+    class Creator:
+        def gen_object_ref(self, data="test", owner=None):
+            return ray.put(data, _owner=owner)
+
+    @ray.remote(resources=actor_resources["owner"], num_cpus=0)
+    class Owner:
+        def __init__(self):
+            self.ref = None
+
+        def set_object_ref(self, ref):
+            self.ref = ref
+
+        def warmup(self):
+            return 0
+
+    @ray.remote(resources=actor_resources["borrower"], num_cpus=0)
+    class Borrower:
+        def get_object(self, ref):
+            return ray.get(ref)
+
+    owner = Owner.remote()
+    creator = Creator.remote()
+    borrower = Borrower.remote()
+
+    # Make sure the owner actor is alive.
+    ray.get(owner.warmup.remote())
+
+    object_ref = creator.gen_object_ref.remote(data="test1", owner=owner)
```

Comment:
I'm a bit confused, isn't the creator ref (`object_ref` above) sufficient to keep the object alive? It's fine for the object to get deleted if the ref count drops to zero.

Reply:
Cause `creator` is not the object's owner, so when `creator` returns this object to `borrower`, the `creator` will don't tell `owner`  and if `creator` got killed,  the `owner` will think all borrower die.

In this pytest case, only two type reference can make this object to keep alive, owner ref and creator ref.

Reply:
Hmm so what's the right solution here, doesn't it work normally?

Cc @stephanie-wang

Reply:
Yeah it should be possible to support in the future, but we'll have to think about what it means for the ref counting protocol.

Reply:
Hmm actually I think this is just one of the limitations of borrowing in the current ref counting protocol. Right now if borrower A creates borrower B and dies before it can report borrower B to the owner, the owner won't know about borrower B.

This is a known edge case in the ref counting protocol, but I don't think it was that relevant until now.

Reply:
I think we can fix this in a followup PR by having borrower A respond early to the WaitForRefRemoved call (which reports borrower B).